test/liblink/socket/recvmsg/recv_state_test.exs

```diff
@@ -0,0 +1,149 @@
+  alias Liblink.Socket.Recvmsg.InitState
+  alias Liblink.Socket.Recvmsg.RecvState
+  alias Liblink.Socket.Recvmsg.SubsState
+  alias Liblink.Socket.Recvmsg.TermState
+
+  @moduletag capture_log: true
+
+  setup do
+    device = %Device{}
+    {InitState, data} = Fsm.new()
+    {:cont, :ok, {RecvState, data}} = InitState.attach(device, data)
+
+    {:ok, [data: data, device: device]}
+  end
+
+  test "can't attach", %{data: data} do
+    assert {:cont, {:error, :badstate}, {RecvState, _}} = RecvState.attach(%Device{}, data)
+  end
+
+  test "at first there is no messages", %{data: data} do
+    assert {:cont, {:error, :empty}, {RecvState, _}} = RecvState.recvmsg(data)
+  end
+
+  test "can halt", %{data: data} do
+    assert {:halt, :ok, {TermState, _}} = RecvState.halt(data)
+  end
+
+  test "recvmsg when queue is empty", %{data: data} do
+    assert {:cont, {:error, :empty}, {RecvState, _}} = RecvState.recvmsg(data)
+  end
+
+  test "recvmsg when queue is not empty", %{data: data} do
+    data = Map.update!(data, :mqueue, &:queue.in([:message], &1))
+    assert {:cont, {:ok, [:message]}, {RecvState, n_data}} = RecvState.recvmsg(data)
+    assert {:cont, {:error, :empty}, {RecvState, _}} = RecvState.recvmsg(n_data)
+  end
+
+  test "recvmsg reverses message", %{data: data} do
+    assert {:cont, :ok, {RecvState, n_data}} = RecvState.on_liblink_message([:bar, :foo], data)
+    assert {:cont, {:ok, [:foo, :bar]}, {RecvState, _}} = RecvState.recvmsg(n_data)
+  end
+
+  test "poll notify right away if there is pending messages", %{data: data} do
+    assert {:cont, :ok, {RecvState, n_data}} = RecvState.on_liblink_message([:foobar], data)
+    assert {:cont, {:ok, tag}, {RecvState, _}} = RecvState.poll(self(), n_data)
+    assert_receive {^tag, :data}
+  end
+
+  test "poll don't notify on_liblink_message if there is pending messages", %{data: data} do
+    assert {:cont, :ok, {RecvState, n_data}} = RecvState.on_liblink_message([:foobar], data)
+    assert {:cont, {:ok, tag}, {RecvState, n_data}} = RecvState.poll(self(), n_data)
+    assert {:cont, :ok, {RecvState, _}} = RecvState.on_liblink_message([:foobar], n_data)
+
+    assert_receive {^tag, :data}
+    refute_receive _
+  end
+
+  test "poll notify on on_libliblink_message", %{data: data} do
+    assert {:cont, {:ok, tag}, {RecvState, n0_data}} = RecvState.poll(self(), data)
+    refute_receive _
+
+    assert {:cont, :ok, {RecvState, _}} = RecvState.on_liblink_message([:message], n0_data)
+    assert_receive {^tag, :data}
+  end
+
+  test "poll notify multiple pollers", %{data: data} do
+    assert {:cont, {:ok, tag0}, {RecvState, n_data}} = RecvState.poll(self(), data)
+    assert {:cont, {:ok, tag1}, {RecvState, n_data}} = RecvState.poll(self(), n_data)
+
+    assert {:cont, :ok, {RecvState, _}} = RecvState.on_liblink_message([:message], n_data)
+    assert_receive {^tag0, :data}
+    assert_receive {^tag1, :data}
+  end
+
+  test "poll is canceled after notification", %{data: data} do
```

Comment:
Este teste não está testando a mesma coisa que o `poll don't notify on_liblink_message if there is pending messages`?

Reply:
Melhor dizendo: aquele teste está testando o mesmo que este?